x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/AuditTrailService.java

```diff
@@ -16,24 +18,40 @@
 import org.elasticsearch.xpack.security.transport.filter.SecurityIpFilterRule;
 
 import java.net.InetAddress;
+import java.time.Duration;
+import java.time.Instant;
 import java.util.Collections;
 import java.util.List;
 
 public class AuditTrailService {
 
+    private static final Logger logger = LogManager.getLogger(AuditTrailService.class);
+
     private static final AuditTrail NOOP_AUDIT_TRAIL = new NoopAuditTrail();
     private final CompositeAuditTrail compositeAuditTrail;
     private final XPackLicenseState licenseState;
+    private final Duration minLogPeriod = Duration.ofMinutes(30);
+    // protected for tests
+    protected Instant lastLogInstant = Instant.EPOCH;
```

Comment:
Without `volatile` means we could end up logging more frequently than every 30 mins from different threads. Do we care?

Reply:
Hmmm, yeah, in a loaded server there could be many groups of such logs.

I think we should strive to keep reads of the instant from a thread local cache and reach to the main memory only in the case that logging is due. But I ended up using an `AtomicReference` after trying with a volatile **in addition** with the regular instance variable.

I'm undecided but I think the atomic reference is good enough.

Reply:
> I think we should strive to keep reads of the instant from a thread local cache and reach to the main memory only in the case that logging is due.
> 
I agree this would probably be more efficient. But I also agree the `AtomicReference` is OK. Because this is a warning situation and ideally users should take action to rectify the warnings. Otherwise, it is OK if things could run a tiny bit inefficiently in this non-ideal situation.

Reply:
Btw, the use of `AtomicReference` does not fully prevent "logging more than once every 30 mins", nor does using `volatile` either. They do reduce the chance of it happening and we seem to be content with its effect, e.g. both `ApiKeyService` and `TokenService` has a `volatile` variable to track last expiration run. So I don't think we need to resort to stronger guarantees (e.g. rw lock) here.

Reply:
Please scratch the above comment, I just noticed the logging call is wrapped inside a `compareAndSet`. This should solve the problem of "no more than once per 30 mins".